src/read/pe/resource.rs

```diff
@@ -0,0 +1,527 @@
+            .field("table", &self.table)
+            .field("entries", &self.entries)
+            .finish()
+    }
+}
+
+impl<'data> ResourceDirectoryTable<'data> {
+    /// Returns the root resource directory, if any.
+    pub fn parse<Pe, R>(pe_file: &PeFile<'data, Pe, R>) -> Result<Self>
+    where
+        Pe: ImageNtHeaders,
+        R: ReadRef<'data>,
+    {
+        let dir = pe_file
+            .data_directory(crate::pe::IMAGE_DIRECTORY_ENTRY_RESOURCE as _)
+            .read_error("No resource directory")?;
+        let data = dir.data(pe_file.data(), &pe_file.section_table())?;
+        Self::read(data, 0)
+    }
+
+    fn read(data: &'data [u8], offset: u64) -> Result<Self> {
+        let table = data
+            .read_at::<ImageResourceDirectory>(offset as _)
+            .read_error("Invalid resource directory table")?;
+        let entries_count = table.number_of_id_entries.get(LE) as usize
+            + table.number_of_named_entries.get(LE) as usize;
+        let offset = offset
+            .checked_add(core::mem::size_of::<ImageResourceDirectory>() as u64)
+            .read_error("Offset overflow")?;
+        let entries = data
+            .read_slice_at::<ImageResourceDirectoryEntry>(offset, entries_count)
+            .read_error("Invalid resource directory entries")?;
+        Ok(Self {
+            directory_data: data,
+            table,
+            entries,
+        })
+    }
+
+    /// Returns an iterator over the directory entries
+    pub fn iter(&self) -> ResourceDirectoryIter<'data> {
+        self.into_iter()
+    }
+
+    /// Returns an iterator over the directory entries
+    pub fn into_iter(self) -> ResourceDirectoryIter<'data> {
+        ResourceDirectoryIter {
+            directory_data: self.directory_data,
+            inner: self.entries.iter(),
+        }
+    }
+}
+
+/// An iterator over a resource directory entries
+#[allow(missing_debug_implementations)]
+pub struct ResourceDirectoryIter<'data> {
+    directory_data: &'data [u8],
+    inner: core::slice::Iter<'data, ImageResourceDirectoryEntry>,
+}
+
+impl<'data> Iterator for ResourceDirectoryIter<'data> {
+    type Item = ResourceDirectoryEntry<'data>;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next().map(|entry| {
+            if entry.is_directory() {
+                ResourceDirectoryEntry::Directory(ResourceDirectory {
+                    directory_data: self.directory_data,
+                    entry,
+                })
+            } else {
+                ResourceDirectoryEntry::Entry(ResourceEntry {
+                    directory_data: self.directory_data,
+                    entry,
+                })
+            }
+        })
+    }
+}
+
+/// A resource directory entry
+#[derive(Debug, Clone, Copy)]
+pub enum ResourceDirectoryEntry<'data> {
+    /// A sub directory entry
+    Directory(ResourceDirectory<'data>),
+    /// A resource entry
+    Entry(ResourceEntry<'data>),
+}
```

Comment:
I think this can be simplified a bit. I want to see something like this instead:
```rust
pub enum ResourceDirectoryEntry<'data> {
    /// A sub directory table.
    Table(ResourceDirectoryTable<'data>),
    /// A resource data entry.
    Data(ResourceDataEntry<'data>),
}
```

I'm not sure exactly though. Feel free to experiment, or I can do further changes after this PR lands.

Reply:
Two things about this:

1. Reading the information about the sub-directory / entry can fails. So the iterator should either:
    - return a simple `Result<_>`, in which case we can not even read the name of the buggy entry.
    - return something more complex which  contains the result and the directory entry.
2. To be able to read the name, we need the `ImageResourceDirectoryEntry`, which you seem to skip in your proposal (it is not in `ResourceDirectoryTable<'data>` at least), so maybe more something like:

```rust
pub enum ResourceDirectoryEntry<'data> {
    /// A sub directory table.
    Table(&'data ImageResourceDirectoryEntry, ResourceDirectoryTable<'data>),
    /// A resource data entry.
    Data(&'data ImageResourceDirectoryEntry,  &'data ImageResourceDataEntry),
}
``` 

Or with the error handling.
```rust
pub enum ResourceDirectoryEntry<'data> {
    /// A sub directory table.
    Table(&'data ImageResourceDirectoryEntry, Result<ResourceDirectoryTable<'data>>),
    /// A resource data entry.
    Data(&'data ImageResourceDirectoryEntry,  Result<&'data ImageResourceDataEntry>),
}
``` 

Don't known, I need to test what's feel cleaner.

Reply:
Don't put the error handling in the enum.

I was thinking that `ResourceDirectoryTable::iter` would return a struct containing the `&'data ImageResourceDirectoryEntry` and probably a flag to indicate if uses an id or name. That struct has a method to return the name. It also has a method to parse whatever is pointed to by the `offset_to_data_or_directory`, and this would return a `Result<ResourceDirectoryEntry>`. I'm not sure on the naming of these (`ResourceDirectoryEntry` might be better for the struct containing the `&'data ImageResourceDirectoryEntry`).

Reply:
Done